File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.js
const room = new Map();

function solution(k, room_number) {
    
    const answer = [];
    
    for(let i=0; i<room_number.length; i++){
        answer.push(findRoom(room_number[i]));
    }
    
    return answer;
}

function findRoom(num){
    
    if(!room.has(num)){
        room.set(num, num+1);
        return num;
    }
    
    var next = findRoom(room.get(num));
    room.set(num, next+1);
    return next;
    
}

File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.java
import java.util.*;

class Solution {
    
    static HashMap<Long, Long> room = new HashMap<>();
    
    public long[] solution(long k, long[] room_number) {
        
        long[] answer = new long[room_number.length];
        
        for(int i=0; i<room_number.length; i++){
            answer[i] = findRoom(room_number[i]);
        }
        
        return answer;
    }
    
    public long findRoom(long number){
        
        if(!room.containsKey(number)){
            room.put(number, number+1);
            return number;
        }
        
        long next = findRoom(room.get(number));
        room.put(number, next);
        return next;
        
    }
}